### ISSUE TITLE
(fea)/omni-opt: Phase 3 - The Unconstrained Descent Engine

### ISSUE THREAD

**Issue by sunhaoxiangwang** [SUM-INNOVATION/rust-scientific-framework#2]
* **Goal:** Implement the primary optimization algorithms for smooth, unconstrained mathematical functions.
* **Scope:** * Steepest Descent.
  * Exact Newton's Method.
  * Levenberg-Marquardt damping modification.
  * Dense BFGS.
  * L-BFGS.
* **Acceptance Criteria:**
  * Steepest descent successfully wires up the gradient ingestion, line search, and stopping criteria as a pipeline validation.
  * Exact Newton delegates to `faer::linalg::cholesky` for positive-definite validation. If factorization fails, the solver gracefully triggers the LM damping parameter $\mu$.
  * Dense BFGS strictly enforces the curvature safeguard, skipping the update if $s^\top y \le \epsilon$.
  * L-BFGS successfully swaps the dense matrix for the `LBFGSWorkspace` and executes the two-loop recursion to compute the descent direction in $O(mn)$ time.
* **Implementation Details:**
  * [cite_start]Newton update: $x^{k+1} = x^k - (D^2f(x^k))^{-1}\nabla f(x^k)$[cite: 30].
  * [cite_start]Levenberg-Marquardt modification: $x^{k+1} = x^k - (D^2f(x^k) + \mu I)^{-1}\nabla f(x^k)$[cite: 32].
  * BFGS must execute the dense rank-2 inverse update.

**Comment by sunhaoxiangwang** [comment]
# Closing Issue #2 — Phase 3: The Unconstrained Descent Engine

## Summary of Achievement

Phase 3 delivers the driver that ties Phases 1 and 2 into a complete unconstrained optimizer. Two new modules were added (`src/methods.rs`, `src/solver.rs`) and `src/oracle.rs` was extended with a feature-gated Hessian trait. The crate now ships **four descent algorithms** behind a single fluent builder:

- **`SteepestDescent`** — pipeline-validation baseline, `d = −∇f(x)`.
- **`BFGS`** — dense rank-2 inverse-Hessian with strict curvature safeguard and γ-scaled warm-start.
- **`LBFGS`** — two-loop recursion driving directly against the Phase-1 ring buffer.
- **`Newton`** (gated on `faer`) — exact Cholesky solve with Levenberg–Marquardt damping fallback for non-PSD Hessians.

### Zero-allocation contract — upheld end-to-end

This engine will power `RUMUS` deep-learning training loops and `OmniNode` peer-to-peer inference, so the zero-allocation contract established in Phase 1 remains absolute. The full audit of the Phase-3 pipeline:

| Site | Allocations | When |
|---|---|---|
| `SolverWorkspace::new` — `x, x_new, g, g_new, d` | 5 × `Vec<f64>` of length `n` | once, in builder |
| Phase-2 `LineSearchWorkspace::new` | 2 × `Vec<f64>` of length `n` | once, in builder |
| `BFGSState::new` — `H, s, y, Hy` | 1 × `n²` + 3 × `n` | once, in builder |
| `LBFGSState::new` → Phase-1 `LBFGSWorkspace::new` | 4 (as per Phase 1) | once, in builder |
| `NewtonState::new` — `h, h_damped, rhs, mem_buf` | 3 × `faer::Mat` + 1 × `MemBuffer` | once, in builder |
| `Solver::run` / `run_with_hessian` — hot path | **0** | per iteration |

Per-iteration mechanics:

- The iterate / gradient boundary at the end of each iteration is a pair of `std::mem::swap` calls — `O(1)` pointer rotations, no memcpy.
- L-BFGS touches the ring directly (`head_s_mut`, `head_y_mut`, `advance`), consuming the Phase-1 workspace unchanged — **no Phase-1 amendment was necessary**.
- Newton's Cholesky runs `cholesky_in_place` + `solve_in_place` against a pre-sized `MemStack`, so even the LM retry loop is allocation-free.
- BFGS rank-2 update and the two-loop recursion use only the buffers allocated in their respective `new`.
- `sᵀy` is computed in a single pass from the live `(x, x_new, g, g_new)` vectors before committing any write — the L-BFGS update never materializes dedicated `s`/`y` scratch.

## Mathematical Foundations

### Exact Newton step

Strictly as directed:

$$
x^{k+1} = x^{k} - \bigl(\nabla^2 f(x^{k})\bigr)^{-1}\,\nabla f(x^{k}).
$$

Implemented via in-place Cholesky: `L L^\top = \nabla^2 f(x^k)`, then `solve_in_place` delivers the descent direction.

### Levenberg–Marquardt damping fallback

When `\nabla^2 f(x^{k})` is not positive definite, `faer`'s Cholesky returns `LltError::NonPositivePivot`; the solver retries with monotonically increasing damping `\mu`:

$$
x^{k+1} = x^{k} - \bigl(\nabla^2 f(x^{k}) + \mu I\bigr)^{-1}\,\nabla f(x^{k}).
$$

`\mu` starts at `initial_mu` (default `10^{-4}`), grows by `mu_growth` (default `\times 10`) after each failed factorization, and bails out with `MethodError::HessianNotFactorable` → `TerminationStatus::NumericalFailure` once `\mu > \text{max\_mu}`.

### BFGS inverse-Hessian update

With the usual conventions `s_k = x_{k+1} - x_k`, `y_k = \nabla f_{k+1} - \nabla f_k`, and `\rho_k = 1 / (s_k^\top y_k)`:

$$
H_{k+1} = \bigl(I - \rho_k\, s_k y_k^{\top}\bigr)\,H_k\,\bigl(I - \rho_k\, y_k s_k^{\top}\bigr) + \rho_k\, s_k s_k^{\top}.
$$

### Curvature safeguard

The rank-2 update (dense BFGS) and the ring push (L-BFGS) are both skipped when

$$
s_k^{\top} y_k \le \varepsilon,
$$

with `\varepsilon \ge 0` supplied by the user via `Method::BFGS { curvature_epsilon }` / `Method::LBFGS { curvature_epsilon }`. `0` is legal (skip only non-positive pairs); larger values reject tiny-curvature pairs that would destabilize `H_k`.

### γ-scaled warm-start for dense BFGS

On the **first** accepted `(s, y)` pair, `H` is reset from the identity to `\gamma I` with

$$
\gamma = \frac{s^{\top} y}{y^{\top} y},
$$

before the rank-2 update is applied — classical BFGS scaling that keeps the first direction in the right ballpark on badly-scaled problems.

### L-BFGS two-loop recursion

At iteration `k` with history depth `m`:

$$
\begin{aligned}
q &\leftarrow \nabla f(x_k);\\
\text{for } i &= k-1, k-2, \dots, k-m:\\
&\qquad \alpha_i = \rho_i\, s_i^{\top} q;\quad q \leftarrow q - \alpha_i\, y_i;\\
r &\leftarrow \gamma_k\, q, \quad\text{where } \gamma_k = \frac{s_{k-1}^{\top} y_{k-1}}{y_{k-1}^{\top} y_{k-1}};\\
\text{for } i &= k-m, \dots, k-1:\\
&\qquad \beta = \rho_i\, y_i^{\top} r;\quad r \leftarrow r + (\alpha_i - \beta)\, s_i;\\
d_k &\leftarrow -r.
\end{aligned}
$$

Implemented with `d` as the sole working vector (`q` in the first loop, `r` in the second, negated at the end) — no scratch beyond the ring's own `alpha` buffer.

## Code Citations

| File | New / changed surface |
|---|---|
| [`omni-opt/src/oracle.rs`](omni-opt/src/oracle.rs) | `HessianOracle` trait (gated on `faer`) |
| [`omni-opt/src/methods.rs`](omni-opt/src/methods.rs) | `Method`, `MethodError`, internal `MethodState`, `BFGSState`, `LBFGSState`, `NewtonState` |
| [`omni-opt/src/solver.rs`](omni-opt/src/solver.rs) | `SolverBuilder`, `Solver`, `SolverWorkspace` (`pub(crate)`), `SolverReport`, `DirectionStep` dispatch trait, `NoHessianStep` / `HessianStep` tokens |
| [`omni-opt/src/lib.rs`](omni-opt/src/lib.rs) | module wiring + crate-root re-exports (`Method`, `MethodError`, `Solver`, `SolverBuilder`, `SolverReport`, gated `HessianOracle`) |

Design decisions locked by the Director during review:

- **Twin entry points `run` / `run_with_hessian`** gated by trait bound (`Oracle` vs `HessianOracle`) — Rust's type system enforces that Newton-selecting users provide a Hessian oracle.
- **Swap-based iterate boundary** via `std::mem::swap` — `O(1)` pointer rotation, zero copy.
- **Armijo extra gradient evaluation accepted and documented** — `\nabla f(x_{k+1})` is refreshed for the next direction.
- **Separate `h_damped` buffer** for Newton/LM — the raw analytic Hessian stays available for diagnostics.
- **BFGS initial `H_0 = I`**, switched to `\gamma I` on first good pair before the rank-2 update.
- **`SolverReport::iters` is total iterations only** — ring flushes / damping triggers deferred to a future diagnostic layer.

## Hardening Notes

Codex issued a FAIL during review because:

1. `curvature_epsilon` on `BFGSState` and `LBFGSState` lacked a boundary check; a `NaN` / `-∞` / negative value would silently poison every update decision.
2. `SolverWorkspace` was `pub`, letting a caller construct a `Solver` around an unvalidated workspace and bypass the builder's tolerance / Wolfe-constant checks.

Both were fixed:

### `curvature_epsilon` guard (both states)

```rust
assert!(
    curvature_epsilon.is_finite() && curvature_epsilon >= 0.0,
    "BFGSState::new: curvature_epsilon must be finite and non-negative (got {})",
    curvature_epsilon
);
```

(identical message shape in `LBFGSState::new`). Four new `#[should_panic(expected = "…")]` regression tests cover NaN, `±∞`, and negative inputs for both states.

### Workspace visibility

```rust
pub(crate) struct SolverWorkspace { /* fields */ }

impl SolverWorkspace {
    pub(crate) fn new(n: usize, method: Method) -> Self { /* … */ }
}
```

The type and its constructor are no longer reachable outside the crate, making `SolverBuilder::build` the only path to a validated `Solver`. Users now *cannot* skip the `0 < c_1 < c_2 < 1` check or the non-zero-dimension check.

## Verification

- `cargo test` — **54 passed / 0 failed**.
- `cargo test --features faer` — **57 passed / 0 failed** (3 Newton / LM tests + 4 hardening regressions).
- `cargo clippy --all-features --all-targets -- -D warnings` — clean.
- Phase 1 and Phase 2 invariants re-verified unchanged; no amendment was required to `LBFGSWorkspace`.

---

**Issue #2 is closed.** The engine is ready to drive `RUMUS` training and `OmniNode` inference. Phase 4 begins on approval.
